### PR TITLE
[no_early_kickoff] Add a special head node resource and use it to pin the serve controller to the head node

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
We used to use `NodeAffinitySchedulingStrategy` to pin the serve controller to the head node but it doesn't work with head node restart since, after restart, head node has a new id. This PR introduces a system label that's automatically added to the head node and use that label to pin the controller to the head node.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/node.py
Comment (edoakes) [inline python/ray/_private/node.py:464]:
```suggestion
                        " is a reserved resource name, use another name instead."
```

**Review by edoakes — COMMENTED** [review]
LGTM.

Let me do some manual testing on this PR before we merge.

**Comment by edoakes** [comment]
Note from testing: let's filter this resource out from user-facing resources in dashboard and autoscaler status.

File: python/ray/autoscaler/_private/util.py
Comment (jjyao) [reply python/ray/autoscaler/_private/util.py:608]: yes, this fixed dashboard as well.

File: python/ray/autoscaler/_private/util.py
Comment (ericl) [inline python/ray/autoscaler/_private/util.py:605]: This makes me think we should use a name like `node:head` to be more consistent with other resources?

**Review by ericl — COMMENTED** [review]
How about let's change this to `node:head` or `node:__head__`?

File: python/ray/autoscaler/_private/util.py
Comment (jjyao) [reply python/ray/autoscaler/_private/util.py:605]: Good point.

**Comment by edoakes** [comment]
Hmm I think these tests won't run on the majority of Serve PRs... is it possible to move this test case into `serve/tests` or is that difficult for some reason?

**Review by ericl — APPROVED** [review]

**Review by edoakes — APPROVED** [review]
OK moving test to `serve/` in follow-up

**Review by pcmoritz — APPROVED** [review]
I'm ok with this, but let's make clear somehow that this is a developer API.

When we finish the labels REP, it should be moved to labels, and we probably should review which labels we provide out of the box (like this one) before we make it a stable API.

**Comment by ollie-iterators** [comment]
The lint test would be fixed by merging in recent changes from the main branch. It was specifically fixed (most recently) by: https://github.com/ray-project/ray/pull/36037

**Pull request merged**
